Add Annotation component

derekr opened the pull request:
Adds `Annotation` component for rendering custom React elements as the annotation content.

```tsx
<Map>
  <Annotation latitude={46.52} longitude={6.57}>
    <div>Some content</div>
  </Annotation>
</Map>
```

Not sure if this is inline with how you'd like to support annotations. I just had a use case for rendering custom markers and thought I'd open a PR upstream. Happy to iterate on this if you're down!

Nicolapps reviewed (changes requested):
Thank you very much for your contribution! It is great to have this feature in mapkit-react.

I remarked a strange bug where the annotation contents aren’t centered anymore if the position changes after it is initially rendered. I don’t have the time right now to investigate more, but it is probably linked to the current implementation where a new `div` wrapper element is created every time `latitude` or `longitude` changes.

<video src="https://user-images.githubusercontent.com/7029582/216509571-40c2c2b8-ecda-4904-b674-ea641686452f.mov"></video>

Also, could you please update [`support.md`](https://github.com/Nicolapps/mapkit-react/blob/main/support.md) in your PR to reflect the changes?

derekr commented on package.json:
```diff
@@ -17,7 +17,8 @@
     "storybook": "start-storybook -p 6006",
-    "build-storybook": "build-storybook"
+    "build-storybook": "build-storybook",
+    "prepare": "npm run build"
```
I just needed this for my fork to build in my own project. Was planning on removing it if you were OK w/ the PR direction, but can def move to `prepublishOnly` if it's helpful for your workflow.

derekr commented on src/components/Annotation.tsx:
```diff
@@ -0,0 +1,79 @@
+    if (map === null) return undefined;
+
+    const el = document.createElement('div');
+    el.id = 'annotation-content';
```
Ah yes! Thanks. This was only for debugging purposes.

derekr commented:
Thank you for the feedback! I think I addressed everything you mentioned, but not sure if I updated support.md completely. Let me know if you think of anything else.

I confirmed the annotation position bug should be resolved. Good catch!

Nicolapps commented on src/components/Annotation.tsx:
```diff
@@ -0,0 +1,78 @@
+    }, [annotation, handler]);
+  });
+
+  return contentRef.current && createPortal(children, contentRef.current);
```
Same thing as above, the check here isn’t necessary.
```suggestion
  return createPortal(children, contentRef.current);
```

Nicolapps commented on src/components/Annotation.tsx:
```diff
@@ -0,0 +1,78 @@
+
+  // Coordinates
+  useEffect(() => {
+    if (map === null) return undefined;
+    if (!contentRef.current) return undefined;
```
This check isn’t necessary since `contentRef.current` always contains an object, you can remove it.

```suggestion
    if (map === null) return undefined;
```

Nicolapps commented on src/components/Annotation.tsx:
```diff
@@ -0,0 +1,78 @@
+    return () => {
+      map.removeAnnotation(a);
+    };
+  }, [map, contentRef, latitude, longitude]);
```
`contentRef` shouldn’t be a dependency of `useEffect`, since `useRef` [will always return the same object](https://beta.reactjs.org/reference/react/useRef#returns).

```suggestion
  }, [map, latitude, longitude]);
```

Nicolapps commented on src/components/Annotation.tsx:
```diff
@@ -0,0 +1,78 @@
+  children,
+}: AnnotationProps) {
+  const [annotation, setAnnotation] = useState<mapkit.Annotation | null>(null);
+  const contentRef = useRef<HTMLDivElement>(document.createElement('div'));
```
This is a little bit inefficient since it will call `document.createElement` on every render (even if the result will be ignored after the first render). Since we never mutate the object, we could use `useMemo` instead, which will do what we want. What do you think about this alternative?

derekr replied:
Really helpful! Ref didn't feel quite right, but I totally forgot about `useMemo`.

Nicolapps reviewed (changes requested):
Thank you very much for the changes! I left a few comments about your use of `useRef`. I can merge your PR after they are fixed.

Nicolapps reviewed (approved):
Thank you very much!
